### ISSUE TITLE
Consider adding a `#[diplomat::constructor]` attribute

### ISSUE THREAD

**Issue by QnnOkabayashi** [rust-diplomat/diplomat#166]
Unlike Rust, languages like C++ and JS support constructors, which are the idiomatic way to construct objects in those languages. We should consider adding a `#[diplomat::constructor]` attribute macro that supports this. For example, we could write
```rust
#[diplomat::opaque]
pub struct Employee {
    id: u64,
}

impl Employee {
    #[diplomat::constructor]
    pub fn new(id: u64) -> Box<Self> { ... }
}
```
Which could generate something like
```js
class Employee {
    constructor(id) { ... }
}
```
... and so we could create objects like this
```ts
const employee = Employee(4);
```
The main issue I foresee with this is that if we have multiple methods for building a type, they would all need to call the constructor under the hood in the autogenerated code, which would be expecting type arguments for the method with the attribute. This might be in issue in JS, but in C++ we can do constructor overloading and have a "raw" constructor to use in other functions.

**Comment by Manishearth** [comment]
+1

So for language-specific stuff, _in general_ i've wanted to err towards something like `#[diplomat::tool_attr(cpp, constructor)]`, i.e. each tool can define its own set of attrs that it will listen to. I think that framework will be useful for a lot of such features.

Constructors .... I could go either way, with something backend specific or something general.

As for C++ we can also have an explicit, private, raw constructor for that stuff internally.

**Comment by sffc** [comment]
I think constructors are a common-enough feature that it deserves a top-level attribute.

However, the applicability to ICU4X types is limited. We're going to have a lot of examples where there are multiple different constructors depending on the types of the arguments. There are not many types that have a single constructor.

I see a few outcomes:

1. The constructor attribute is only for the default constructor (no arguments).
2. We "pick" one function to make the constructor for each class, and that function gets special status.
3. We allow the attribute to annotate multiple functions, and languages can choose to make overloading constructors.

Option 3 seems like the most flexible moving forward, but it should maybe be part of a larger issue about allowing functions to have overloads.

**Comment by Manishearth** [comment]
Worth noting that Dart autochooses a constructor (I believe in some cases Dart must have a ctor, too. I'm not sure of this).

An attribute for explicit ctor promotion would be great.

Strawman proposal: `#[diplomat::constructor]` that takes in an optional `retain_original = false` parameter, depending on whether you want to keep the method around with the "normal" name as well.

**Comment by robertbastian** [comment]
> Worth noting that Dart autochooses a constructor (I believe in some cases Dart must have a ctor, too. I'm not sure of this).

Dart always requires a constructor, but it can be private. The current Dart behaviour is purely ergonomic.

**Comment by Manishearth** [comment]
Fixed in #447
